pkg/controller/controller.go

```diff
@@ -1165,11 +1189,17 @@ func isFinalError(err error) bool {
 		return false
 	}
 	switch st.Code() {
+	case codes.ResourceExhausted: // CSI: operation not pending, "Unable to provision in `accessible_topology`"
+		if mayReschedule {
+			// may succeed elsewhere -> give up for now
+			return true
+		}
+		// may still succeed at a later time -> continue
+		return false
```

Comment:
This is unfortunate case of mixing gRPC transport error codes with application / CSI errors.

*Go gRPC implementation* returns `ResourceExhausted` when it processes too big messages (both when sending and receiving). This is the logic behind the current code - the operation itself may be in progress or even successfully finished (too big success response may have been just thrown away) and the provisioner should retry.

Looking at the code, gRPC default max message size is [sometimes defaulted to 4KiB](https://github.com/kubernetes-csi/external-provisioner/blob/e73400e556e73c0c43448fc497aefe26b6f2dc4f/vendor/google.golang.org/grpc/server.go#L56)

When a *CSI driver* intentionally returns `ResourceExhausted` when it gets out of space on the storage backend, the error is final and we can be 100% sure that the volume is being provisioned. Rescheduling the pod / PVC to a different node is correct.

Questions are:
* How to recognize *who* sent the response. IMO it's impossible to distinguish CSI driver from gRPC implementation.
* Whether we can assume that `ResourceExhausted` returned by gRPC implementation is safe **not** to retry. 
  * If CreateVolumeResponse / CreateVolumeRequest was too big, retrying the request should yield the same response and one of them is going to be too big again. Retrying is useless and there is a leaked volume. Trying on a different node won't harm.

Reply:
This is a problem in the CSI spec, right? Instead of re-using status codes, it would have been better to pick new ones to avoid such ambiguity.

I agree that it is impossible to determine the sender. If sending the request already fails, then trying elsewhere is safe and in fact the only way how (theoretically) the volume might still succeed, because retrying locally will just send the same message over and over again.

When receiving the reply fails with `ResourceExhausted`, then the situation is worse in the sense that the volume probably has been provisioned, external-provisioner just never gets to know that. Retrying won't help here much either unless an admin steps in an reconfigures the external-provisioner with larger max message sizes.

I think we should bring this up as a problem with spec. But given what it is right now, avoiding the gRPC `ResourceExhausted` by using large enough message size limits and then assuming that the error really comes from the CSI driver seems viable to me.